sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collationExpressions.scala

```diff
@@ -73,24 +75,57 @@ object CollateExpressionBuilder extends ExpressionBuilder {
  * This function is pass-through, it will not modify the input data.
  * Only type metadata will be updated.
  */
-case class Collate(child: Expression, collationName: String)
-  extends UnaryExpression with ExpectsInputTypes {
-  private val collationId = CollationFactory.collationNameToId(collationName)
-  override def dataType: DataType = StringType(collationId)
+case class Collate(child: Expression, collation: Expression)
+  extends BinaryExpression with ExpectsInputTypes with CodegenFallback {
```

Comment:
why it needs to be `CodegenFallback` fallback now? It should delegate codegen to `child`

Reply:
Because if we call `defineCodeGen` then it will implicitly call CodeGen on left and right child which results in execution error (right child is unevaluable).

```scala
  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
    defineCodeGen(ctx, ev, (childGen, _) => childGen) // exception, generates both left and right code
```
or
```scala
  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
    left.doGenCode(ctx, ev) // forbiden
```

Reply:
We can follow what `Alias` does
```
  /** Just a simple passthrough for code generation. */
  override def genCode(ctx: CodegenContext): ExprCode = child.genCode(ctx)
  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
    throw SparkException.internalError("Alias.doGenCode should not be called.")
  }
```